proposals/webidl-bindings/Explainer.md

```diff
@@ -0,0 +1,647 @@
+implementation) go through the [Web IDL ECMAScript Binding] layer.  This is
+enabled by allowing a WebAssembly module to include a list of bindings for
+imports and exports that specify how to create a Web IDL function from a
+WebAssembly function and vice versa.
+
+An import binding describes how to create a WebAssembly [host function instance][`hostfunc`]
+instance from a Web IDL function by specifying:
+* the Web IDL function signature, including parameter and result types and
+  whether the Web IDL function is a static (the current default), method (with
+  an additional `this` parameter) or constructor (with an additional `newTarget`
+  parameter).
+* for each Web IDL parameter, how to construct the Web IDL argument value out of
+  one or more WebAssembly values and possibly the contents of linear memory.
+* for the Web IDL return value, how to destructure it into one or more
+  WebAssembly values and/or the contents of linear memory.
+
+Conversely, an export binding describes how to create a Web IDL [Callback] value
+from a WebAssembly function by specifying the same kinds of things as an import
+binding, but "in reverse".
+
+One important property of the proposal is that there can be many bindings for a
+given Web IDL signature.  This is important for several reasons:
+* As new WebAssembly value types are added (e.g., with [Function References] and
+	[GC]), they can be bound in new ways to Web IDL signatures which would not be
+  possible if, e.g., a string always had to be passed via an `i32` offset
+  into linear memory.
+* Even with the WebAssembly MVP's set of types, there can be multiple ways to
+  produce a given Web IDL value (e.g., null-terminated vs. explicit length 
+  vs. `anyref` for a string).
+
+The set of features described in this proposal is likely too large for an
+initial release and therefore will likely be broken down into an initial "MVP"
+feature set. Moreover, additional bindings for non-functions are out of scope
+of this proposal but could be proposed in the future. For example:
+* Bindings could be added mapping Web IDL [Exceptions] to the [Exception Tags]
+  of the WebAssembly [Exception Handling] proposal, enabling WebAssembly to
+  throw and catch Web IDL exceptions.
+* Bindings could be added for [Type Exports] to allow WebAssembly modules to
+  create [ES Classes] enabling, e.g., Web Components to be fully implemented in
+  WebAssembly.
+
+Given this overview, the proposal is now presented in a bottom-up fashion.
+
+
+## Web IDL Types
+
+Symmetric to the core WebAssembly [Type Section], all compound Web IDL types are
+defined in a **Web IDL Type Subsection** that is stored inside of, and can be
+referenced throughout the rest of, the new **Web IDL Bindings Section** being
+defined by this proposal. The section/subsection design of Web IDL Bindings
+would mirror that of the [Name Section] and allow for extension in the future.
+
+A tentative list of Web IDL compound types includes:
```

Comment:
i don't think we need to make this restriction. The difference between this subset and the whole of webIDL is not that significant; and, I may be misunderstanding, but using Type Imports to extend the supported webIDL sounds like an uneccessary complication.

Reply:
There's actually quite a lot that is outside the subset covered here; see also @littledan's comments specifically about this above too.  Focusing on a subset lets us work incrementally.  Moreover, we have to declare the types anyway, so I don't see any strong motivation to attempt to cover the entirety of Web IDL.

Type Imports don't *extend* Web IDL, they provide static type checking within a wasm module of existing Web IDL types, avoiding the dynamic downcast from `any` on every call.